doc/specification/language-proposal.md

```diff
@@ -0,0 +1,1518 @@
+
+Example:
+
+```
+let a = { scale = 3.2; pos = { x = 3 m ; y = 4 m; }; };   # binds `a` below.
+with a.pos;   # binds `x` to 3 m and `y` to 4 m below.
+a.scale*(x+y);
+```
+
+As the type of a value binding is the type of its final expression, syntactic ambiguity in type assertion parsing has no semantic consequence: `let id = expr₁; expr₂: T` has the type of `expr₂: T`, which is `T` iff the type of `expr₂` is `T` or a supertype of `T` ; similarly, `(let id = expr₁; expr₂): T` has type `T` iff the type of `let id = expr₁; expr₂`, which is the type of `expr₂` is `T` or a supertype of `T`.
+
+
+### Conditional expressions
+<a id="conditional-expressions"/>
+
+> _conditional-expr_ ::= _if-expr_ | _case-expr_
+>
+> _if-expr_ ::= `if` _boolean-expr_ `then` _expression_ `else` _expression_
+>
+> _case-expr_ ::= `|` _boolean-expr_ ***right-arrow*** _expression_ )? _case-expr_ | `|` ( `otherwise` | `true` ) ***right-arrow*** _expression_
+
+Alternatives based on one or more conditions can be expressed with if/then/else clauses or case expressions introduced with the bar symbol. In either, the conditional expression is ill-formed if any of the _boolean-expr_ have type which is not boolean.
+
+The value of `if condition then expr₁ else expr₂` is the value of `expr₁` in the case where `condition` is true, and `expr₂` otherwise. The type of the _if-expr_ is the type of `expr₁` and `expr₂`; the expression is ill-formed if the types of `expr₁` and `expr₂` differ.
+
+The value of `| otherwise → expr` is just the value of `expr`. `true` can be used in place of `otherwise` equivalently. For a compound _case-expr_ of the form `| condition → expr₁ case₂`, the value is `expr₁` in the case where `condition` is true, and the value of `case₂` otherwise; the expression is ill-formed if the types of `expr₁` and `case₂` differ.
+
+Example:
+
+```
+# Equivalent to 10 km:
+let a = if 3>2 then 10 m else 2 m; a*1000
+
+# Piece-wise constant reaction rate function.
+def rate = fn (T: temperature) →
+  | T <  0 °C → 10 mM/s
+  | T < 10 °C → 20 mM/s
+  | otherwise → 30 mM/s;
+
+```
+
+As with value bindings, there is no semantic ambiguity arising from an ambiguous parse from a trailing type assertion.
+
+
+### Boolean expressions
+<a id="boolean-expressions"/>
+
+> _boolean-expr_ ::= _boolean-term_ ( `or` _boolean-term_ )*
+>
+> _boolean-term_ ::= _boolean-factor_ ( `and` _boolean-factor_ )*
+>
+> _boolean-factor_ ::= `not`? ( _comparison-expr_ | _function-application_ | _qualified-identifier_ | _boolean-literal_ | `(` _boolean-expr_ `)` )
+>
+> _comparison-expr_ ::= _comparison-base_ _comparison-op_ _comparison-base_
+>
+> _comparison-op_ ::= ***compare-equal*** | ***compare-not-equal*** | ***compare-less*** | ***compare-less-equal*** | ***compare-greater*** | ***compare-greater-equal***
+>
+> _comparison-base_ ::= _algebraic-expr_ | `(` _expression_ `)`
```

Comment:
It's mainly there to force the precedence, so that
```
(let a = 3; a) < 3   # result of let expression less than 3?
let a = 3; a < 3     # let expression has value 3 < 3.
```
They necessarily both work out the same here, but without the syntax above, the latter has an ambiguous parse as the former.

It also exists to impose some order on expressions like the following:
```
let a = true; a == false == false; # ill-formed
let a = true; (a == false) == false; # equivalent to (true == false) == false
```